framework/src/new_operators/mod.rs

```diff
@@ -114,21 +147,65 @@ pub mod tests {
         use packets::udp::tests::UDP_PACKET;
 
         dpdk_test! {
-            let mut batch = PacketBatch::new(&UDP_PACKET).map(|p| p.parse::<Ethernet>()).for_each(
-                |p| {
-
-                    println!("I'm a Side Effect");
-                    let x = 1;
-                    let y = x;
-                    println!("{}", x+y);
-                    println!("Ethernet is a packet ~ {}", p);
+            let mut invoked = 0;
+
+            {
+                let (producer, batch) = single_threaded_batch::<RawPacket>(1);
+                let mut batch = batch.for_each(|_| {
+                    invoked += 1;
                     Ok(())
-                }
-            );
+                });
+                producer.enqueue(RawPacket::from_bytes(&UDP_PACKET).unwrap());
 
-            let packet = batch.next().unwrap().unwrap();
-            assert_eq!(EtherTypes::Ipv4, packet.ether_type())
+                let _ = batch.next();
+            }
+
+            assert_eq!(1, invoked);
         }
     }
 
+    #[test]
+    fn group_by_operator() {
+        use packets::tcp::tests::TCP_PACKET;
+        use packets::udp::tests::UDP_PACKET;
+
+        dpdk_test! {
+            let (producer, batch) = single_threaded_batch::<RawPacket>(2);
+
+            let mut batch = batch
+                .map(|p| p.parse::<Ethernet>()?.parse::<Ipv4>())
+                .group_by(
+                    2,
+                    |p| match p.protocol() {
+                        ProtocolNumbers::Tcp => 0,
+                        ProtocolNumbers::Udp => 1,
+                        _ => 2
+                    },
+                    |mut groups| {
+                        merge![
+                            groups.remove(&0)
```

Comment:
I think we discussed wrapping these up as they seem odd in the API/interface sense.

Reply:
know you're on it @drunkirishcoder